source/simulator/assembly.cc

```diff
@@ -1551,6 +1636,906 @@ namespace aspect
+                                       (neighbor_density_c_P + neighbor_latent_heat_LHS)
+                                       * time_step
+                                       * (current_u
+                                          * scratch.subface_finite_element_values.normal_vector(q))
+                                       * scratch.neighbor_face_phi_field[i]
+                                       * scratch.face_phi_field[j]
+                                       :
+                                       0.)
+
+                                   )
+                                   * scratch.subface_finite_element_values.JxW(q);
+
+                                data.local_matrices_ext_ext[f * GeometryInfo<dim>::max_children_per_face + subface_no](i,j)
+                                += (+ 0.5 * time_step * neighbor_conductivity
+                                    * scratch.neighbor_face_grad_phi_field[i]
+                                    * scratch.subface_finite_element_values.normal_vector(q)
+                                    * scratch.neighbor_face_phi_field[j]
+
+                                    + 0.5 * time_step * neighbor_conductivity
+                                    * scratch.neighbor_face_grad_phi_field[j]
+                                    * scratch.subface_finite_element_values.normal_vector(q)
+                                    * scratch.neighbor_face_phi_field[i]
+
+                                    + time_step
+                                    * max_density_c_P_and_latent_heat
+                                    * max_penalty
+                                    * scratch.neighbor_face_phi_field[i]
+                                    * scratch.neighbor_face_phi_field[j]
+
+                                    + (!inflow
+                                       ?
+                                       (neighbor_density_c_P + neighbor_latent_heat_LHS)
+                                       * time_step
+                                       * (current_u
+                                          * scratch.subface_finite_element_values.normal_vector(q))
+                                       * scratch.neighbor_face_phi_field[i]
+                                       * scratch.neighbor_face_phi_field[j]
+                                       :
+                                       0.)
+
+                                   )
+                                   * scratch.subface_finite_element_values.JxW(q);
+                              }
+                          }
+
+                      }
+                  }
+              }
+            /**
+             * else there are no subfaces, so we are either on a subface
+             * of the neighbor, or we share the whole face with the neighbor.
+             */
+            else
+              {
+                if (!cell->neighbor_is_coarser(f) &&
+                    (cell->subdomain_id() < neighbor->subdomain_id() ||
```

Comment:
I would move this `else-if` above the `if` in line 1902 since, conceptually, it is the simpler case to treat.

Also, you know (and can test via an assertion) that `cell` is locally owned, so I think the condition here is not quite correct. You need to assemble *always* if the neighbor is a ghost cell, because you're never going to see that neighbor in `cell`. If you do this, then statements such as `cell->subdomain_id() == neighbor->subdomain_id()` may in fact be easier to state as `neighbor->is_locally_owned()`.